File: docs-examples/example-http-function-java/src/test/java/example/BookControllerTest.java
package example;

import io.micronaut.core.type.Argument;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.client.RxHttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.http.client.exceptions.HttpClientResponseException;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;

import jakarta.inject.Inject;
import java.util.List;

import static io.micronaut.http.HttpStatus.BAD_REQUEST;
import static io.micronaut.http.HttpStatus.CREATED;
import static io.micronaut.http.HttpStatus.OK;
import static io.micronaut.http.MediaType.APPLICATION_JSON_TYPE;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

@MicronautTest
public class BookControllerTest {

    @Inject
    @Client("/")
    RxHttpClient client;

    @Test
    void testValidation() {
        HttpClientResponseException e = assertThrows(HttpClientResponseException.class, () -> client
                .exchange(HttpRequest.POST("/books", new Book("", 400))
                        .contentType(APPLICATION_JSON_TYPE), Book.class)
                .blockingFirst()
        );

        assertEquals(BAD_REQUEST, e.getResponse().status());
    }

    @Test
    void testListBooks() {
        final HttpResponse<Book> postBookResponse = client
                .exchange(HttpRequest.POST("/books", new Book("Along Came a Spider", 400))
                        .contentType(APPLICATION_JSON_TYPE), Book.class)
                .blockingFirst();

        assertEquals(CREATED, postBookResponse.status());

        assertNotNull(postBookResponse.body());
        assertEquals(400, postBookResponse.body().getPages());

        HttpResponse<List<Book>> response = client
                .exchange(HttpRequest.GET("/books"), Argument.listOf(Book.class))
                .blockingFirst();

        assertEquals(OK, response.status());

        final List<Book> body = response.body();
        assertNotNull(body);
        assertEquals(2, body.size());
    }
}

File: docs-examples/example-http-function-python/src/test/python/example/BookControllerTest.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.core.type import Argument
from micronaut.http import HttpRequest, HttpStatus, MediaType
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.http.client.exceptions import HttpClientResponseException
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test
from reactor.core.publisher import Mono

from example.Book import Book


@MicronautTest
class BookControllerTest:
    client: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def test_validation(self) -> None:
        try:
            Mono.from_(self.client.exchange(
                HttpRequest.POST("/books", Book("", 400)).contentType(MediaType.APPLICATION_JSON_TYPE),
                Book)).block()
        except HttpClientResponseException as e:
            assert e.getResponse().status() == HttpStatus.BAD_REQUEST
        else:
            assert False, "validation should have failed"

    @Test
    def test_list_books(self) -> None:
        post_book_response = Mono.from_(self.client.exchange(
            HttpRequest.POST("/books", Book("Along Came a Spider", 400)).contentType(MediaType.APPLICATION_JSON_TYPE),
            Book)).block()

        assert post_book_response.status() == HttpStatus.CREATED

        assert post_book_response.body() is not None
        assert post_book_response.body().pages == 400

        response = Mono.from_(self.client.exchange(HttpRequest.GET("/books"), Argument.listOf(Book))).block()

        assert response.status() == HttpStatus.OK

        body = response.body()
        assert body is not None
        assert len(body) == 2
